### DISCUSSION TITLE
On "extract_dynamic_coords" function

### DISCUSSION THREAD

**Discussion by Crizzo98** [r-a-dobson/dynamicSDM#12] [Q&A]
Hello everyone!
I am testing the dSDM package for a project of mine and have a question about using the "extract_dynamic_coords" function.
I am extracting data from a monthly dataset (TERRACLIMATE, link: https://developers.google.com/earth-engine/datasets/catalog/IDAHO_EPSCOR_TERRACLIMATE) and I have daily presence/pseudoab occurrences data. When I use the function "extract_dynamic_coords" I can't omit the arguments "GEE.math.fun" and "temporal.res". All I need is the single climatic data for my occurences data (e.g. maximum temperature for that month), not an average or any other function that I can calculate with "GEE.math.fun". I don't need, for example, the sum of the precipitation for x days before my data, as in the tutorials.
How should I set these two arguments?
Thank you!

**Comment by r-a-dobson** [comment]
Hi there,

Thank you for getting in touch about using the extract_dynamic_coords function.

The function works by collecting all the bands from a dataset within a temporal window and then processing them. For your specific case, I would recommend setting the temporal resolution as 31 days prior and using the GEE.math.fun argument as "last". This will return the value of the most recent band in the dataset from the month prior to your occurrence data.

By setting GEE.math.fun to "last", you are essentially using the Google Earth Engine reducer ee.Reducer.last(), which returns the last value of its inputs (https://developers.google.com/earth-engine/apidocs/ee-reducer-last). 

Best wishes,
Rachel

**Reply by r-a-dobson** [reply]
(i.e., if it is a monthly dataset, there should only really be one band, but "last" just ensures you get the most recent monthly value available prior to your occurrence).

**Reply by Crizzo98** [reply]
Thank you so much!

**Comment by FereshtehNamdari** [comment]
Hi Rachel,

I am trying to use the DynamicSDM package for my project. Before applying it to my own dataset, I wanted to test it with the dataset you shared.

However, I encountered a problem at the extract_dynamic_coords function— it took a long time to run because of different errors and eventually for this error:

Error in value[[3L]](cond): 
  It looks like your EE credential has expired. Try running ee_Authenticate() again or clean your credentials ee_clean_user_credentials()
  
I saw in one of the GitHub issues that this error could be solved by running:

(https://github.com/r-spatial/rgee/issues/369) rgee::ee$Initialize(project = "Your_Project") 
I tried that, using the name of a project I had already created and uploaded my CSV file to, and it worked.


Now I have two questions:

Do I need to upload your CSV file to the project that appears in https://code.earthengine.google.com/ > Assets → Cloud Assets to make it work?

Or does it simply work through the project path, regardless of what is uploaded there, and just use the CSV file I read from R?

Because when I specify a new project name that is empty , it still doesn’t work.

Thank you !

**Comment by randle-cp** [comment]
Hello: 
I've been trying to work through the tutorial for extracting explanatory variables for two days and am having persisten problems with authentication. When I run the extract dynamic coords command
```
extract_dynamic_coords(occ.data=sample_filt_data,
                       datasetname = "UCSB-CHG/CHIRPS/DAILY",
                       bandname="precipitation",
                       spatial.res.metres = 5566 ,
                       GEE.math.fun = "sum",
                       temporal.direction = "prior",
                       temporal.res = 56,
                       save.method = "split",
                       varname = variablenames[1],
                       save.directory = extraction_directory_1)
```
I get the following result
```
◉  Python version
✔ [Ok] /Users/SHSU/.virtualenvs/rgee/bin/python v3.12
◉  Python packages:
✔ [Ok] numpy
✔ [Ok] earthengine-api
NOTE: The Earth Engine Python API version 1.7.28 is installed
correctly in the system but rgee was tested using the version
1.6.11. To avoid possible issues, we recommend install the
version used by rgee (1.6.11). You might use:
* rgee::ee_install_upgrade()
* reticulate::py_install('earthengine-api==1.6.11', envname='PUT_HERE_YOUR_PYENV')
* pip install earthengine-api==1.6.11 (Linux and Mac0S)
* conda install earthengine-api==1.6.11 (Linux, Mac0S, and Windows)
◉  Credentials neccesaries for rgee:
✔ [Ok] Earth Engine Credentials found.
✔ [Ok] Google Drive credentials found.
✔ [Ok] Google Cloud Storage credentials found.
── rgee 1.1.8 ────────────────────────────────────────────── earthengine-api 1.7.28 ── 
 ✔ user: not_defined 
 ✔ Initializing Google Earth Engine:
Warning in py_call_impl(callable, call_args$unnamed, call_args$named) :
ee.ee_exception.EEException: Not signed up for Earth Engine or project is not registered. Visit https://developers.google.com/earth-engine/guides/access
Run `reticulate::py_last_error()` for details.
✔ Initializing Google Earth Engine: DONE!
Error in value[[3L]](cond) :
It looks like your EE credential has expired. Try running ee_Authenticate() again or clean your credentials ee_clean_user_credentials().
```
So I use `ee_Authenticate()` and `ee_clean_user_credentials` but get the same error message.  I realize that this is likely a probelm with rgee but I love the tools you offer in this package and would just like to know if there is a good workaround?
I use macOS 15.7.7, with R version 4.6.0
I have much more experience in R than python, so maybe that''s what I'm missing?
Thank you so much.
